docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,372 @@
+  conv.i[0] = 0;
+
+  return conv.d;
+}
+```
+
+* Comment the non-obvious, not the obvious.
+
+``` c++
+// Turns out sorting this list before processing it is faster due to branch
+// prediction, so don't remove this without checking the profile:
+std::sort(users.begin(), users.end(), UserSortPredicate);
+for (User const &user : users)
+{
+  if (user.ready)
+  {
+    // Do something...
+  }
+}
+```
+
+* If you do not understand some code, contact the original author. If they have left the company ask the most competent person. After working out what a piece of a non-obvious code is doing, remember to document it.
+
+
+## Naming
+
+* Use meaningful names, even in tests.
+* For struct and class names, use `PascalCase`.
+* Private and protected class member variables end with an `_`.
+* Variable names are `snake_case_` with trailing underscore, except in structs where it is `snake_case` without underscore.
+* For templates, use all uppercase with underscores to separate words. Long and explicit names are preferred to short and obscure names, e.g. `template <typename ARRAY_TYPE>` is preferable to `template <typename A>);`. Short names are acceptable where the type is obvious and reused many times in complex equations.
+* For enums use `UPPER_CASE`.
+* For macros use `UPPER_CASE`.
+
+
+## Formatting
+
+* Make use of the `apply_style` script provided for formatting.
+* Lines should be no longer than 100 characters.
+* Use 2 spaces to indent, not tabs.
+* Curly brackets `{...}` should be on their own line.
+* There should be a space after conditionals, i.e. `if (test)`
+* No spaces inside `(parenthesis)` or `<template>` unless it improves readability.
+* No spaces around arrow: `this->fnc();`.
+* No trailing whitespace.
+* Align attributes by group in classes, and initialisers.
+* Variables should be initialised with `=` if primitive, or braced initialisation with parentheses if required.
+* Initialiser lists should be initialised with braces by default.
+* Use SQL style initialiser for lists and inheritance lists.
+
+```c++
+Foo::Foo()
+  : attribute1_{value1}
+  , attribute2_{value2}
+{}
+```
+
+
+## Typing
+
+* Use STL types wherever possible.
+* Prefer `(u)int(16/32/64)_t` and `std::size_t`.
+* Use `static_cast`, `reinterpret_cast`, `dynamic_cast`, etc., rather than C-style casts.
+* Avoid `auto` except when the return type is clear.
+* Avoid use of RTTI.
+
+* `using` is preferable to `typedef` for type alias definition.
+* Make `using` private unless you intend to expose it.
+* `namespace {.<template T>` only allowed if `typedef`/`using` indicates the type.
+
+* If a value is `const`, declare it so.
+* Use `int const &i` over `const int &i`.
+
+* Prefer `enum class` over regular `enum`.
+
+## Namespaces
+
+* Do not do `using namespace foo` or `namespace baz = ::foo::bar::baz` in `.hpp` files.
+* Do not do `using namespace foo` in the top `fetch` namespace, as it pollutes all of it. Instead, do the following:
+
+```c++
+namespace fetch
```

Comment:
Again the unindented version on that PR is consistent with the apply style script